pkg/build/controller/factory/factory.go

```diff
@@ -261,6 +260,45 @@ func (factory *ImageChangeControllerFactory) Create() controller.RunnableControl
 	}
 }
 
+type BuildConfigControllerFactory struct {
+	Client                  osclient.Interface
+	BuildConfigInstantiator buildclient.BuildConfigInstantiator
+	// Stop may be set to allow controllers created by this factory to be terminated.
+	Stop <-chan struct{}
+}
+
+// Create creates a new ConfigChangeController which is used to trigger builds on creation
+func (factory *BuildConfigControllerFactory) Create() controller.RunnableController {
+	queue := cache.NewFIFO(cache.MetaNamespaceKeyFunc)
+	cache.NewReflector(&buildConfigLW{client: factory.Client}, &buildapi.BuildConfig{}, queue, 2*time.Minute).RunUntil(factory.Stop)
+
+	bcController := &buildcontroller.BuildConfigController{
+		BuildConfigInstantiator: factory.BuildConfigInstantiator,
+	}
+
+	return &controller.RetryController{
+		Queue: queue,
+		RetryManager: controller.NewQueueRetryManager(
+			queue,
+			cache.MetaNamespaceKeyFunc,
+			func(obj interface{}, err error, retries controller.Retry) bool {
+				bc := obj.(*buildapi.BuildConfig)
+				if maxRetries > retries.Count {
+					glog.V(3).Infof("Giving up retrying BuildConfig update event %s/%s: %v", bc.Namespace, bc.Name, err)
+					kutil.HandleError(err)
+					return false
+				}
+				glog.V(4).Infof("Retrying BuildConfig update event %s/%s: %v", bc.Namespace, bc.Name, err)
+				return true
```

Comment:
in case of a collision event here, aren't you going to end up retrying N times before actually giving up on the event?

HandleBuildConfig should differentiate the errors it returns so that if the instantiate failure is because something else already kicked off the "first build", the event doesn't get retried here.